<a href="https://colab.research.google.com/github/caio-emiliano/desafio_solvimm/blob/main/desafio_engenharia_dados_solvimm.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Análise dos Dados 5GFlix

Para o estudo de mercado do aplicativo 5GFlix, nosso objetivo será responder algumas questões e obter *insights* através dos dados fornecidos pelo *database*.

Para isso, deveremos ser capazes de responder as seguintes perguntas:



1.   Quantos filmes estão disponíveis no dataset?
2.   Qual é o nome dos 5 filmes com melhor média de avaliação?
3.   Quais os 5 anos com menos lançamentos de filmes?
4.   Quantos filmes que possuem avaliação maior ou igual a 4.7, considerando apenas os filmes avaliados na última data de avaliação do dataset?
5.  Dos filmes encontrados na questão anterior, quais são os 10 filmes com as piores notas e quais as notas?
6.  Quais os id's dos 5 customer que mais avaliaram filmes e quantas avaliações cada um fez?

Obs: Este notebook se resolve sozinho. Deste que as bibliotecas importadas estejam instaladas, o código rodará sem que seja necessário criar pastas ou editar arquivos.

In [ ]:
#Importar as bibliotecas necessárias para o código

import pandas as pd

In [ ]:
#Importar os arquivos CSVs deste cenário. Upei os arquivos no Google Drive e os leio via link para que o código seja executável por qualquer um (apenas por simplicidade)

movies_path = "https://www.googleapis.com/drive/v3/files/1tenHUdbV7Zy2ggauVAK-6PzHHIrbYcR8?alt=media&key=AIzaSyDbIgqySdMFMP085-zVCVPNHFIoJpuhVRo"
df_movies = pd.read_csv(movies_path, sep=';', names=['movie_id', 'name-year']) #Leitura do arquivo csv. O parâmetro sep é para indicar que o delimitador deste csv é o ';'. O parâmetro names indica o nome das colunas 
                                                                                                                                                                            #que serão criadas no dataset

customersRating_path = "https://www.googleapis.com/drive/v3/files/1jYxQa6gGzUksidziQWLTAou3KXwFlPvy?alt=media&key=AIzaSyDbIgqySdMFMP085-zVCVPNHFIoJpuhVRo"
df_customers = pd.read_csv(customersRating_path, sep=';') #Leitura do arquivo csv. O parâmetro sep é para indicar que o delimitador deste csv é o ';'.

In [ ]:
df_movies

movie_id                                          name-year
0            1                            (Dinosaur Planet, 2003)
1            2                 (Isle of Man TT 2004 Review, 2004)
2            3                                  (Character, 1997)
3            4               (Paula Abdul's Get Up & Dance, 1994)
4            5                   (The Rise and Fall of ECW, 2004)
...        ...                                                ...
4494      4495  (Clifford: Happy Birthday Clifford / Puppy Lov...
4495      4496                      (Farewell My Concubine, 1993)
4496      4497                                 (Texasville, 1990)
4497      4498                                      (Gonin, 2000)
4498      4499                                 (In My Skin, 2002)

[4499 rows x 2 columns]

Se observarmos a base de dados acima, perceberemos que nome e ano encontram-se juntos. Precisaremos separar essa informação e dividi-la em 2 colunas.

Para isso, usaremos o comando split para separarmos na ocorrência de vírgula, e criaremos 2 novas colunas para o dataframe receber essa informação.

In [ ]:
df_movies[['name', 'year']] = df_movies['name-year'].str.split(',', n=1, expand=True)
df_movies

movie_id  ...    year
0            1  ...   2003)
1            2  ...   2004)
2            3  ...   1997)
3            4  ...   1994)
4            5  ...   2004)
...        ...  ...     ...
4494      4495  ...   2002)
4495      4496  ...   1993)
4496      4497  ...   1990)
4497      4498  ...   2000)
4498      4499  ...   2002)

[4499 rows x 4 columns]

Agora que os dados estão formatados corretamente, podemos excluir a coluna name-year.

In [ ]:
df_movies.drop('name-year', axis=1, inplace=True)

###Análise Exploratória dos dados

Neste cenário teremos 2 datasets diferentes. O primeiro - df_movies - contém o id de cada filme, nome e ano de lançamento. Já o segundo dataset - df_customers - contém o Id do cliente que fez a avaliação, a nota de avaliação, a data da avaliação feita e o Id do filme avaliado.

Assim, irei analisar abaixo as dimensões dos dataframes.

In [ ]:
#Dimensões dos Dados:
print("DIMENSÕES DO DATAFRAME MOVIES:")
print("Linhas:\t\t{}".format(df_movies.shape[0]))
print("Colunas:\t{}".format(df_movies.shape[1]), "\n")

print("DIMENSÕES DO DATAFRAME CUSTOMERS:")
print("Linhas:\t\t{}".format(df_customers.shape[0]))
print("Colunas:\t{}".format(df_customers.shape[1]))

DIMENSÕES DO DATAFRAME MOVIES:
Linhas:		4499
Colunas:	3 

DIMENSÕES DO DATAFRAME CUSTOMERS:
Linhas:		24053764
Colunas:	4


Ao olharmos para os dados, podemos perceber que cada linha do dataframe Movie representa um filme diferente. Logo, como a base de dados possui 4499 linhas **podemos responder a pergunta nº 1: existem 4499 filmes no dataset.**

In [ ]:
df_movies

movie_id                                             name    year
0            1                                 (Dinosaur Planet   2003)
1            2                      (Isle of Man TT 2004 Review   2004)
2            3                                       (Character   1997)
3            4                    (Paula Abdul's Get Up & Dance   1994)
4            5                        (The Rise and Fall of ECW   2004)
...        ...                                              ...     ...
4494      4495  (Clifford: Happy Birthday Clifford / Puppy Love   2002)
4495      4496                           (Farewell My Concubine   1993)
4496      4497                                      (Texasville   1990)
4497      4498                                           (Gonin   2000)
4498      4499                                      (In My Skin   2002)

[4499 rows x 3 columns]

**Para respondermos a pergunta nº 2**, primeiramente precisamos obter a média de avaliação de cada filme. Para isso, irei agrupar as informações em df_customers por filme e então realizar a média dos valores.

Após feita a média, ordenarei os valores em ordem decrescente de média de avaliação. Assim os 5 primeiros valores representarão os filmes com maior média de avaliação.

Por fim, filtraremos a tabela por id do filme e buscaremos o nome no dataset df_movie.

**O resultado me retornará o nome dos 5 filmes com maior média de avaliação**

In [ ]:
ratingMean = df_customers.groupby('Movie_Id').mean().reset_index() #Agrupo o dataset por filme (comando groupby) e faço a média de avaliação de cada filme (mean).

ratingMean.drop('Cust_Id', axis=1, inplace=True) #Como realizamos a média, não sentido esta ser calculado nos Ids dos clientes. Portanto, podemos eliminar essa coluna do dataset

bestRatingMovies = ratingMean.sort_values(by='Rating', ascending=False).head()['Movie_Id'] #Ordeno a tabela por ordem decrescente de média, assim os 5 primeiros valores serão os filmes com a maior média de avaliação. Como estamos interessados no id desses filmes, filtro pela coluna movie_id

df_movies[df_movies['movie_id'].isin(bestRatingMovies)]

movie_id                                               name    year
12          13  (Lord of the Rings: The Return of the King: Ex...   2003)
2101      2102                            (The Simpsons: Season 6   1994)
3032      3033  (Ghost in the Shell: Stand Alone Complex: 2nd Gig   2005)
3455      3456                                    (Lost: Season 1   2004)
4237      4238                                         (Inu-Yasha   2000)

**Para a questão nº 3**, preciso saber a quantidade de filmes lançados por ano. Para isso, agrupei o dataset movies por ano e contei os valores de filmes. Assim, tenho o número de filmes lançados em cada ano.

Depois disso, ordenei os valores em ordem crescente. Logo, **os primeiros 5 resultados são os anos que possuem menos lançamentos**

In [ ]:
movies_byYear = df_movies.groupby('year').count().drop('name', axis=1).rename(columns={'movie_id':'releases'}) #Agrupo a tabela por ano e conto quantos filmes foram lançados em cada ano. Também usei o comando rename para renomear a 
                                                                                                                                                                                   #coluna de filmes contados para release
                                                                                                                                                                                   
movies_byYear.sort_values(by='releases').head() #Ordeno os valores de contagem em ordem crescente e mostro os 5 primeiros (que serão os 5 anos com menos lançamentos)

releases
year            
 1915)         1
 1917)         1
 1922)         1
 1926)         1
 1916)         2

**Para a questão 4**, precisarei saber qual foi a última data de avaliação do Dataset. Para isso, basta ordenarmos os dados por data, em ordem decrescente. Logo, o primeiro resultado será a última data de avaliação.

Feito isso, precisaremos saber todos os filmes com notas maiores ou iguais a 4.7 e que foram avaliados nesta data. 

In [ ]:
df_customers.sort_values(by='Date', ascending=False) #Ordenação do dataset por data em ordem decrescente

Cust_Id  Rating        Date  Movie_Id
19662630   349234     2.0  2005-12-31      3756
9003586   1808587     4.0  2005-12-31      1798
4762203    183702     4.0  2005-12-31       954
22225850  1673594     2.0  2005-12-31      4216
13578622  1922435     4.0  2005-12-31      2590
...           ...     ...         ...       ...
20393918   510180     2.0  1999-11-11      3870
6901473    510180     5.0  1999-11-11      1367
9056171    510180     5.0  1999-11-11      1798
14892677   510180     3.0  1999-11-11      2866
19585852   510180     4.0  1999-11-11      3730

[24053764 rows x 4 columns]

In [ ]:
rated47Movies = df_customers[ df_customers['Rating'] >= 4.7] #Filtro todos os filmes que possuem nota maior ou igual a 4.7

#considerar apenas os avaliados no ultimo dia de avaliação (31/12/2005)
result = rated47Movies[ rated47Movies['Date'] == "2005-12-31"].drop_duplicates(subset="Movie_Id") #removi os dados duplicados para que a contagem (etapa seguinte) fosse feita corretamente

result['Movie_Id'].count() #contar quantos filmes são

780

In [ ]:
result.sort_values(by='Rating', ascending=True) #ordenando os filmes que satisfazem as condições da questao 4 em ordem crescente de nota de avaliação

Cust_Id  Rating        Date  Movie_Id
18187      329442     5.0  2005-12-31         8
15629469   332079     5.0  2005-12-31      3025
15636453  2226028     5.0  2005-12-31      3030
15639905  2005761     5.0  2005-12-31      3033
15650436  1258694     5.0  2005-12-31      3046
...           ...     ...         ...       ...
8139426   1963479     5.0  2005-12-31      1631
8184406   2068826     5.0  2005-12-31      1642
8265688    588193     5.0  2005-12-31      1650
7329244   2254964     5.0  2005-12-31      1470
24043825    23421     5.0  2005-12-31      4496

[780 rows x 4 columns]

Olhando a tabela acima, cheguei a conclusão que todos os filmes com nota maior ou igual a 4.7 e lançados na última data avaliada do dataset (31/12/2005) já possuem nota 5 (já que a ordenação crescente retorna 5 logo no primeiro valor). **Portanto, respondendo as questões 4 e 5, existem 780 filmes e todos estes avaliados com nota 5.**

**Na questão nº 6**, queremos encontrar quantas avaliações cada clientes fez, para então saber os clientes que mais avaliaram filmes. 

Para isso, agrupei os dados por cliente e contei o número de vezes que cada cliente avaliou um filme. Eliminei as colunas 'Date' e 'Movie_Id' para facilitar a visualização da contagem.

Logo em seguida, ordenei os valores por ordem decrescente, de modo que **os resultados do comando head (5 primeiros) são os clientes que mais avaliaram filmes e também resposta da questão 6.**

In [ ]:
numberOfRatingPerClient = df_customers.groupby(by='Cust_Id').count().drop(['Date', 'Movie_Id'], axis=1)

numberOfRatingPerClient.sort_values(by='Rating', ascending=False).head()

Rating
Cust_Id        
305344     4467
387418     4422
2439493    4195
1664010    4019
2118461    3769